### ISSUE TITLE
Lazy vals in type classes lock on the whole instance

### ISSUE THREAD

**Issue by runarorama** [scalaz/scalaz#1144]
E.g. in `Bind.scala`:

``` scala
override def ap[A, B](fa: => F[A])(f: => F[A => B]): F[B] = {
  lazy val fa0 = fa
  bind(f)(map(fa0))
}
```

While `fa` is being evaluated, the evaluating thread occupies the monitor for `this` (the `Bind` instance), and other threads entering `ap` must wait. This is especially egregious for monads like `Future` and `Task` where you might want to have multiple threads calling `ap` at the same time.

Possible fix:

``` scala
override def ap[A, B](fa: => F[A])(f: => F[A => B]): F[B] = {
  val monitor = new Object { lazy val fa0 = fa }
  bind(f)(map(monitor.fa0))
}
```

In this specific case, I'm not sure it's even necessary to cache the `fa` value since `map` is strict. There are other instances of this though. For example, `Monad.whileM`.

**Comment by xuwei-k** [comment]
use `scalaz.Need` instead of `new Object{` ?

**Comment by runarorama** [comment]
Not possible in every case, but yeah, I'm currently going through and replacing `lazy val` with `Need`.

**Comment by jedws** [comment]
I think the title should be changed, the problem is the lazy val, not the by name argument.

**Comment by aloiscochard** [comment]
Nice catch @runarorama, and what a terrible scala gotcha :-(

**Comment by runarorama** [comment]
@jedws The title says "by need", not "by name".

**Comment by alexandru** [comment]
Hi, I noticed this thread on Twitter, sorry for adding my opinion, couldn't help it :-P

Personally I already knew how `lazy val` behaves, because its behavior can't come for free and I checked out the generated bytecode a long time ago. Checking out the generated bytecode is a healthy habit, with the `javap` utility being available and easy to use. And due to the `lazy val` constraints, the current implementation is very reasonable:

1. `lazy val` has guaranteed idempotency in the presence of multi-threading and effects, or in other words the lazy expression will be evaluated only once. And in order to guarantee idempotency in this case, you can't escape the use of a lock. It's impossible. In other instances, like when you've got an idempotent `() => Unit`, you might escape with a simple `getAndSet`, but not this one.
2. it does synchronize on `this`, using what's called the intrinsic lock, but that's because it needs some instance to synchronize on. A `lazy val` could have its own `AnyRef` or monitor to synchronize with, but that would waste memory. Which is acceptable for higher-level contracts in libraries, but not OK for language constructs.
3. `lazy val` does the [double-checked locking](https://en.wikipedia.org/wiki/Double-checked_locking) technique, so once the value has been cached and *seen* then it doesn't do the locking anymore.
4. for lexically scoped / local lazy vals, locking on `this` isn't acceptable, but theoretically the JVM can do *lock elision* by means of [escape analysis](https://docs.oracle.com/javase/7/docs/technotes/guides/vm/performance-enhancements-7.html#escapeAnalysis). Not sure how well that works though, I never measured it.

Personally I avoid usage of `lazy val`. But I think that doing `val monitor = new Object { lazy val fa0 = fa }` is wasteful, especially because of number 4. With the JVM it is never easy to infer its behavior, especially when it comes to intrinsic locks. Personally I'd recommend some benchmarks with JMH, as you might discover that the workaround is worse than the problem.

Cheers,

**Comment by jedws** [comment]
@runarorama ack

**Comment by xuwei-k** [comment]
FYI
- http://docs.scala-lang.org/sips/pending/improved-lazy-val-initialization.html
- https://github.com/lampepfl/dotty/pull/81

**Comment by puffnfresh** [comment]
@alexandru I disagree that the current lazy val implementation is very reasonable but I also think that is irrelevant.

**Comment by alexandru** [comment]
@puffnfresh the intent above isn't to defend lazy val but to warn that the behavior of the JVM isn't intuitive. The Java standard library is a landmine of misapplied synchronization, yet it still works fine usually because of its optimizations. Extra objects on the other hand put stress on the GC and add an extra level of indirection. I wouldn't do that.

**Comment by DarkDimius** [comment]
https://gist.github.com/DarkDimius/03a19bdc5dc099c96508 - this gist demonstrates that behavior is changed in dotty: no syncrhonziation on `this` is done.

Dotty also supports concurrent initialization of independent lazy fields in objects, though compilation scheme there is more complicated.

**Comment by puffnfresh** [comment]
@alexandru what do you recommend?

**Comment by runarorama** [comment]
Sounds to me like @alexandru is volunteering to write a macro

**Comment by runarorama** [comment]
There's a further problem with using a wrapper object, or `Need`:

``` scala
def foo[A,B,C](a: => A, f: (=> A) => B, g: (=> A) => C): (B, C) {
  val an = Need(a)
  (f(a), g(a))
}

def bar[A,B,C](a: => A, f: (=> A) => B, g: (=> A) => C): (B, C) {
  lazy val an = a
  (f(a), g(a))
}
```

In `bar`, the thunk reference `a` is cleared before entering `f` and `g`. In `foo`, a new reference to the thunk is created inside the `Need`, so until `an` goes out of scope, the `Function0[A]` cannot be GCed.

Do we care about this?

**Comment by aloiscochard** [comment]
@runarorama sounds negligible to me
